ghcide/src/Development/IDE/Core/WorkerThread.hs

```diff
@@ -42,13 +45,19 @@ withWorkerQueue workerAction = ContT $ \mainAction -> do
                 workerAction l
 
 -- | 'awaitRunInThread' queues up an 'IO' action to be run by a worker thread,
--- and then blocks until the result is computed.
+-- and then blocks until the result is computed. If the action throws an
+-- non-async exception, it is rethrown in the calling thread.
 awaitRunInThread :: TQueue (IO ()) -> IO result -> IO result
 awaitRunInThread q act = do
     -- Take an action from TQueue, run it and
     -- use barrier to wait for the result
     barrier <- newBarrier
     atomically $ writeTQueue q $ do
-        res <- act
-        signalBarrier barrier res
-    waitBarrier barrier
+        resultOrException <- try act
+        case resultOrException of
+            Left e@(fromException -> Just AsyncCancelled) -> throwIO e
+            _ -> signalBarrier barrier resultOrException
+    resultOrException <- waitBarrier barrier
+    case resultOrException of
+        Left e  -> throwIO (e :: SomeException)
+        Right r -> return r
```

Comment:
This will bubble up to the rules engine? Or how does the logging work here?

Reply:
It will bubble up to the rules engine. And being catch here https://github.com/haskell/haskell-language-server/blob/67680ea347308f8f6819cb6f55c9a60c9bb19313/ghcide/src/Development/IDE/Core/Shake.hs#L1248

Reply:
Just like other kind of rule failure.